src/core/agent_space_interaction.jl

```diff
@@ -185,3 +185,49 @@ function add_agent_single!(model::ABM, properties...)
     return agent
   end
 end
+
+
+####################################
+### sample
+####################################
+
+"""
+        sample!(model::ABM [, weight]; kwargs...)
+
+Replace agents of the `model` with a random sample of the current agents with
+size `n`. Optionally, choose an agent property `weight`(Symbol) to weight the sampling.
+
+# Keywords
+
+* `n`: the number of new agents.
+* `replace=true` whether sampling is performed with replacement.
+* `rng`::AbstractRNG=Random.GLOBAL_RNG: Optionally, specify a random number generator.
+"""
+function sample!(model::ABM, weight=nothing; n::Int, replace=true, ordered=false, rng::AbstractRNG=Random.GLOBAL_RNG)
+    if weight != nothing
+        weights = Weights([getproperty(i, weight) for i in values(model.agents)])
+        newids = sample(rng, collect(keys(model.agents)), weights, n, replace=replace, ordered=ordered)
+    else
+        newids = sample(rng, collect(keys(model.agents)), n, replace=replace, ordered=ordered)
+    end
+
+    newagents = [deepcopy(model.agents[i]) for i in newids]
```

Comment:
Why are the agents deepcopied? In general, what does this line aim to achieve?

Reply:
Because I noticed that without deepcopying, when an agent is replaced more than once, all the resulting new agents are connected: mutating one of them would mutate them all.